streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java

```diff
@@ -277,197 +275,361 @@ public void close() {
+         * @param withTimestampIterator The iterator for the timestamped column family.
+         * @param storeName             The name of the store associated with this iterator.
+         * @param forward               {@code true} for forward iteration; {@code false} for reverse iteration.
+         * @param toInclusive           Whether the upper boundary of the range is inclusive.
+         */
+        RocksDBDualCFRangeIterator(final Bytes from,
+                                   final Bytes to,
+                                   final RocksIterator noTimestampIterator,
+                                   final RocksIterator withTimestampIterator,
+                                   final String storeName,
+                                   final boolean forward,
+                                   final boolean toInclusive) {
             this.forward = forward;
+            this.noTimestampIterator = noTimestampIterator;
+            this.storeName = storeName;
+            this.toInclusive = toInclusive;
+            this.withTimestampIterator = withTimestampIterator;
+
+            this.rawLastKey = initializeIterators(from, to);
         }
 
+        /**
+         * Retrieves the next key-value pair in the range.
+         *
+         * <p>This method determines the next key-value pair to return by merging the results from the two column
+         * families. If both column families have keys, it selects the one that matches the iteration order and range
+         * conditions. Keys outside the specified range are skipped.</p>
+         *
+         * @return The next {@link KeyValue} pair in the range, or {@code null} if no more elements are available.
+         */
         @Override
-        public synchronized boolean hasNext() {
-            if (!open) {
-                throw new InvalidStateStoreException(String.format("RocksDB iterator for store %s has closed", storeName));
-            }
-            return super.hasNext();
+        protected KeyValue<Bytes, byte[]> makeNext() {
+            loadNextKeys();
+            if (noTimestampNext == null && withTimestampNext == null) return allDone();
+            final KeyValue<Bytes, byte[]> next = fetchNextKeyValue();
+            return isInRange(next) ? next : allDone();
+        }
+
+        /**
+         * Returns the next key in the range without advancing the iterator.
+         *
+         * <p>This method retrieves the key of the next {@link KeyValue} pair that would be returned by {@link #next()},
+         * without moving the iterator forward. This is useful for inspecting the next key without affecting the
+         * iterator's state.</p>
+         *
+         * @return The next key as a {@link Bytes} object.
+         *
+         * @throws NoSuchElementException If there are no more elements in the iterator.
+         */
+        @Override
+        public Bytes peekNextKey() {
```

Comment:
existing test coverage for this class is below average (e.g. there's not `peekNextKey` in the test) - could you add that? 

cc @ableegoldman it may be good to refactor the existing tests alongside this one so that instead of one really big test we had individual tests for each piece of functionality so it's easier to confirm the test coverage is good

Reply:
In progress.